### DISCUSSION TITLE
How the ID Token Encryption Content Encryption Algorithm works?

### DISCUSSION THREAD

**Discussion by lrodriguez14** [keycloak/keycloak#9876] [Q&A]
Hello,

I have my keycloak server running and it works pretty well, I have a frontend public client in Angular that is login in and getting the access token, but I want to encrypt it to be more secure, and I am wondering how ID Token Encryption Content Encryption Algorithm works? I can select any algorithm  and when I try to log in, I am getting an error **error_description: can not get encryption KEK**, any idea how to set it up?

**Comment by tnorimat** [comment]
@lrodriguez14 Hello, could you refer to "Signed and Encrypted ID Token Support" section of [Keycloak's Server Administration guide](https://www.keycloak.org/docs/latest/server_admin/index.html#con-advanced-settings_server_administration_guide)?

**Comment by lrodriguez14** [comment]
Thanks Takashi, it was really helpful. Now I am able to encrypt my idtoken but looks like that the JS Adapter is failing when it tries to decode it (Because it is encrypted), is there any way to solve this?

```
ERROR URIError: malformed URI sequence
    decodeToken keycloak.js:1065
    setToken keycloak.js:1004
    authSuccess keycloak.js:806
    onreadystatechange keycloak.js:791
    Angular 8
core.js:6498
    Angular 3
    RxJS 5
    Angular 11
```

Also, once that I have the token encrypted how can I decrypt it?

**Reply by tnorimat** [reply]
@lrodriguez14 Unfortunately, it seems that the current keycloak's JS adapter does not support decrypting JWE encoded token. You might find some other JS library that supports JWE.
